#!/bin/bash

# 1) Extract only multiallelic sites

# zcat variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered.vcf.gz.recode.vcf.gz | grep -v "^##" | awk '$5 ~ ","' | cut -f 1,2 > multiallelic_sites.bed

# 2) Merge vcf files (heterozygous in hybrids and homozygous in nonhybrids) only with the multiallelic sites

bcftools merge -R multiallelic_sites.bed variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_AUS2_B_1_14_merged_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_BF3_B_1_14_merged_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_EML11-2_GTGGCC_L001_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H05_EML11-200_E_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H06_TAN16-2-3C_Y2_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H07_TAN16-2-3C_X3_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H08_EML11-200_I_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H10_EML11-200_H_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H11_TAN16-2-3C_Y1_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H12_KWT3-5_X1_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H14_TAN16-2-3C_X2_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H18_TWN10-2_X7_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H19_EML11-200_X1_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H20_TWN11-1_X1_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H21_EML11-2_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H23_TAN16-2-3C_X1_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_hybrid_H26_EML11-200_G_heter_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_KWT3_B_1_14_merged_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_LO1_B_1_14_merged_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_LO2_B_1_14_ATCACG_L002_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_LO4_B_14_AGTCAA_L002_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_A_TWN2-1_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_B_Aus1-1_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_K_EML11-200_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_S_TWN10-2_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_T_TAN16-2-3C_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_parent_U_KWT3-5_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_PHIL-17_CAGATC_L001_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_TAN02-1_ACTTGA_L001_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_TAN03-3_CTTGTA_L001_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_TW10_B_1_14_merged_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_ZAM03-2_AGTTCC_L001_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_ZAM07-11_ATTCCT_L001_homoz_only.vcf.gz variants_mapping_rna_Aus_genome_all_samples_only_variant_sites_filtered_ZIM03-13_ACTGAT_L001_homoz_only.vcf.gz  > variants_only_multiallelic_sites_heter_and_homoz.vcf

# 3) Traslate to genotypes
bcftools query -H -f '%CHROM %POS  %REF  %ALT [ %TGT]\n' variants_only_multiallelic_sites_heter_and_homoz.vcf > variants_only_multiallelic_sites_heter_and_homoz.txt

# 4) Reformat colnames and genotypes
sed 's/  */\t/g' variants_only_multiallelic_sites_heter_and_homoz.txt | sed "s/\#\t//g" | sed 's/\[.[^]]\]//g' | sed "s/\[[1-9]\]//g" | sed "s/:GT//g" | sed "s/\./-/g" > variants_only_multiallelic_sites_heter_and_homoz_reformatted.txt 

